### PR TITLE
[FIX] {hr,sale}_timesheet: prevent editing so line if validated timesheet

### PR DESCRIPTION
**Pull request by thle-odoo**
Steps to reproduce:
-------------------
- create a product:
    - service
    - based on timesheet
    - create on order: project & task

- create a sale order with this product
- confirm the sale order
- go on the created task
- add timesheets with an employee

- go to timesheet
- validate created timesheets

- go to the project settings
- in invoicing tab, change de sale order item for the employee

Issue:
------
The sale order item linked to the validated timesheets are modified.

This behaviour causes negative side effects.
A validated timesheet should no longer be modified, but the backend logic does, which is not consistent.

Solution:
---------
Make the condition that filters timesheets to update take account of whether the timesheet is validated or not.

opw-3791062

### CODE REVIEW

**Comment by thle-odoo** [comment]
Linked PR: https://github.com/odoo/enterprise/pull/61122

File: addons/sale_timesheet/models/project.py [hunk 1]
```diff
@@ -211,9 +211,13 @@ def write(self, values):
             })
         return res
 
+    def _possibly_updated_timesheets(self):
```
Comment (xavierbol) [inline addons/sale_timesheet/models/project.py:214]: We could name it `_get_timesheets` (we have the same method in `project.task` for instance) or `_get_updatable_timesheets` for instance. 🙂

File: addons/sale_timesheet/models/project.py [hunk 2]
```diff
@@ -211,9 +211,13 @@ def write(self, values):
...
+        self.ensure_one()
+        return self.mapped('timesheet_ids').filtered(lambda t: not t.is_so_line_edited and t._is_not_billed())
+
     def _update_timesheets_sale_line_id(self):
         for project in self.filtered(lambda p: p.allow_billable and p.allow_timesheets):
-            timesheet_ids = project.sudo(False).mapped('timesheet_ids').filtered(lambda t: not t.is_so_line_edited and t._is_not_billed())
+            timesheet_ids = project.sudo(False)._possibly_updated_timesheets()
```
Comment (xavierbol) [inline addons/sale_timesheet/models/project.py:220]: I think the `sudo(False)` is useless here (I know you did not add it, Laurent and I added it but I think it was useless. 😇 )
```suggestion
            timesheet_ids = project._possibly_updated_timesheets()
```

File: addons/sale_timesheet/models/project.py [hunk 3]
```diff
@@ -211,9 +211,13 @@ def write(self, values):
             })
         return res
 
+    def _possibly_updated_timesheets(self):
+        self.ensure_one()
+        return self.mapped('timesheet_ids').filtered(lambda t: not t.is_so_line_edited and t._is_not_billed())
```
Comment (xavierbol) [inline addons/sale_timesheet/models/project.py:216]: Equivalent to:

```suggestion
        return self.timesheet_ids.filtered(lambda t: not t.is_so_line_edited and t._is_not_billed())
```

File: addons/sale_timesheet/models/project.py (hunk 2 above)
Comment (xavierbol) [inline addons/sale_timesheet/models/project.py:220]: Alternative solution is to define a method in `account.analytic.line` model to return True if the timesheet is updatable. 🙂 

Something like this:

```diff
diff --git a/addons/hr_timesheet/models/hr_timesheet.py b/addons/hr_timesheet/models/hr_timesheet.py
index 6a66a2cba5f4..503e332eaf1e 100644
--- a/addons/hr_timesheet/models/hr_timesheet.py
+++ b/addons/hr_timesheet/models/hr_timesheet.py
@@ -351,6 +350,9 @@ class AccountAnalyticLine(models.Model):
         company_uom = self.env.company.timesheet_encode_uom_id
         return company_uom == self.env.ref('uom.product_uom_day')

+    def _is_updatable_timesheet(self):
+        return True
+
     @api.model
     def _convert_hours_to_days(self, time):
         uom_hour = self.env.ref('uom.product_uom_hour')
diff --git a/addons/sale_timesheet/models/account.py b/addons/sale_timesheet/models/account.py
index 5baa0c5daa86..576bdb6299b0 100644
--- a/addons/sale_timesheet/models/account.py
+++ b/addons/sale_timesheet/models/account.py
@@ -174,3 +174,6 @@ class AccountAnalyticLine(models.Model):
     def _timesheet_convert_sol_uom(self, sol, to_unit):
         to_uom = self.env.ref(to_unit)
         return round(sol.product_uom._compute_quantity(sol.product_uom_qty, to_uom, raise_if_failure=False), 2)
+
+    def _is_updatable_timesheet(self):
+        return super()._is_updatable_timesheet() and self._is_not_billed()
```

+ the diff I will share in your enterprise PR.

and then we just have to call inside the `filtered` method. 🙂 

```suggestion
            timesheet_ids = project.mapped('timesheet_ids').filtered(lambda t: not t.is_so_line_edited and t._is_updatable_timesheet())
```

it is up to you to choose the solution you prefer. It is reeally the same for me 🙂

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂

**Review by xavierbol — COMMENTED** [review]
LGTM, thanks for your work. 🙂 

robodoo r+

**Pull request closed without merging**